tests/conftest.py

```diff
@@ -564,17 +564,20 @@ def __init__(self, mid):
         def _async_fire_mqtt_message(topic, payload, qos, retain):
             async_fire_mqtt_message(hass, topic, payload, qos, retain)
             mid = get_mid()
-            mock_client.on_publish(0, 0, mid)
+            if not topic.startswith("no_callback/"):
```

Comment:
I don't think it's a good pattern to have conditional logic in tests. I'd recommend having a separate fixture for this part that we can override per test.

Reply:
That is not going to work I am afraid of.  The reality is that the client will or will not respond with an ACK. To avoid disturbing the the fixture should hold back all call backs. If a separate fixture would be needed it would become almost a copy mqtt_mock and mqtt_client_mock. That is not need to my opinion.

Reply:
Here's an example how to break out the side effect of a mock to a separate fixture:
https://github.com/home-assistant/core/blob/8e39ba387d0fcbd8462fff76da4d64890bc4ec57/tests/components/zwave_js/test_config_flow.py#L77-L91

Reply:
We can return a factory function from the side effect fixture that is called with the mock as parameter, if the side effect needs access to the mock.

Reply:
To be concrete I want want change `mqtt_client_mock` and `mqtt_mock` which is linked to avoid performing a call back. If that is to be don manually I want to have control over what mid is used to the _wait_for_mid call since the order can be random.

Reply:
With your example I am complete lost, unfortunally. Not sure what to how to override.

Reply:
I suggest we leave the testing of the mid out of this PR for now so we don't block it further.

Reply:
I'll commit an extra test to illustrate a stand-a-lone-example without special fixtures bases on mqtt_client_mock. This is not about publishing twice, it is about handling the ACK client response correctly. See the added `test_wait_for_ack` test. (did not change the rest yet)

Reply:
> I suggest we leave the testing of the mid out of this PR for now so we don't block it further.

Have a look at `test_wait_for_ack`. That could replace `test_publish_mid`. `test_handle_mqtt_on_callback` still needs some adjustments, since it leans on the conditional effects.

Reply:
I think we can test this more simple and strictly. I suggest we we don't include the changes or tests for the mid in this PR.

Reply:
Ok, i'll themove the mid stuff and fixture changes for this PR.

Reply:
Done